### DISCUSSION TITLE
Proper Use of route() in Composition API/SFC

### DISCUSSION THREAD

**Discussion by andrewskm** [tighten/ziggy#492] [Q&A] [answered]
I've been using `this.route()` in my Vue code and everything works fine.
Flipping pages over to use the composition API and I was wonder if the proper way to call `route()` is to use `getCurrentInstance()` like so `getCurrentInstance().root.ctx.route()` or is there a better way?

Everything works, just wanted to make sure it's proper.

**Comment by bakerkretzmar** [comment, marked as answer]
Inside your templates you should be able to keep using `route()` normally, and if you're using the `@routes` Blade directive then `route()` should actually also be available globally, which means you can use it directly even in the `setup()` function.

I didn't know about `getCurrentInstance()`, if that works then I think it's totally fine to do that.

Otherwise, you can import `route` in each component and just use it like a regular function, see https://github.com/tighten/ziggy/issues/454#issuecomment-906418423 for some detailed examples. We're planning to provide a `useRoute()` hook eventually to make this easier too.

**Reply by andrewskm** [reply]
Yeah the route still works in the templates as normal. 
I'm not using the blade directive, so it's not defined in the setup.

Importing from `ziggy/dist` seems fine, but needing to call it like `route(name, params, absolute, config)` may be a lil much, so I may stick with the getCurrentInstance for now.

The `useRoute()` hook would be perfect, especially since I use InertiaJS as they have similar hooks for `useForm` and `usePage`.

Thanks for your help!

**Comment by romsar** [comment]
At the moment I use provide/inject vue functionality that works in composition api.

In setup() you can easilly do `const route = inject('route')`.

In app.js or ssr.js `.provide('route', route)`

**Reply by anburocky3** [reply]
I received this `ESLint: 'route' is not defined.(no-undef)` error on the typescript project.

Screenshot below: 
![image](https://user-images.githubusercontent.com/15264938/152645954-8010d2d1-0cf3-4efc-b508-33bc5e2679f1.png)

**Reply by romsar** [reply]
Did you put `const route = inject('route')` before?
Looks like you didn't define route variable.
Also don't forget to put `.provide('route', route)` and app.js & ssr.js.

**Reply by anburocky3** [reply]
> 

I'm using Inertiajs within Laravel app, So i have used `@routes` inside my layout file. So it means, i can use the `route()` without defining in vuejs right?

My PHPStorm IDE is complaining about the unavailability. But the `route()` will be available on runtime. How to avoid this IDE error?

**Reply by romsar** [reply]
> > 
> 
> I'm using Inertiajs within Laravel app, So i have used `@routes` inside my layout file. So it means, i can use the `route()` without defining in vuejs right?
> 
> My PHPStorm IDE is complaining about the unavailability. But the `route()` will be available on runtime. How to avoid this IDE error?

No, that's normal. Phpstorm doesn't know about variables that are defined in blade file.

**Reply by anburocky3** [reply]
> > > 
> > 
> > 
> > I'm using Inertiajs within Laravel app, So i have used `@routes` inside my layout file. So it means, i can use the `route()` without defining in vuejs right?
> > My PHPStorm IDE is complaining about the unavailability. But the `route()` will be available on runtime. How to avoid this IDE error?
> 
> No, that's normal. Phpstorm doesn't know about variables that are defined in blade file.

Mm yea, so any way to hide this? Like ignoring just the `route()`. I can't proceed without clearing this ide error.?

**Reply by andrewskm** [reply]
You have to add "route" to the eslint globals for it to stop complaining.

**Reply by anburocky3** [reply]
> You have to add "route" to the eslint globals for it to stop complaining.

Tried writing these in `.eslintrc`

```js
    "globals": {
        "route": "writable"
    }
```

but still it is complaining? Am i doing it wrong?

**Reply by livevsonline** [reply]
> provide('route', route)

Hey Roman

Can you help me solve my ssr.ts file? I'm using Laravel 9, Inertia SPA with SSR + Typescript, and Vite...

I have a few problems with the ssr file

![image](https://user-images.githubusercontent.com/39704356/179359006-6b598d9b-b065-423a-bb66-ec3fdf70f7f8.png)

**Reply by romsar** [reply]
> > provide('route', route)
> 
> Hey Roman
> 
> Can you help me solve my ssr.ts file? I'm using Laravel 9, Inertia SPA with SSR + Typescript, and Vite...
> 
> I have a few problems with the ssr file
> 
> ![image](https://user-images.githubusercontent.com/39704356/179359006-6b598d9b-b065-423a-bb66-ec3fdf70f7f8.png)

Sure.

```
// import modules...
import { createSSRApp, h } from 'vue';
import { renderToString } from '@vue/server-renderer';
import { createInertiaApp, Head, Link } from '@inertiajs/inertia-vue3';
import createServer from '@inertiajs/server';
import route from 'ziggy';

createServer((page) => createInertiaApp({
  page,
  render: renderToString,
  resolve: name => require(`./pages/${name}`),
  setup({ app, props, plugin }) {
    const Ziggy = props.initialPage.props.ziggy;

    return createSSRApp({
      render: () => h(app, props),
    })
      .use(plugin)
      .mixin({
        methods: {
          route: (name, params, absolute, config = Ziggy) => route(name, params, absolute, config),
        },
      })
      .provide('route', (name, params, absolute, config = Ziggy) => route(name, params, absolute, config))
  },
}));
```

**Reply by livevsonline** [reply]
Thank you for your time Roman, i get this now:

![image](https://user-images.githubusercontent.com/39704356/179359958-34b6efc6-9edd-4e55-b5e3-a417d27440ac.png)


How come we removed the resolvecomponent part?

**Reply by livevsonline** [reply]
> > > provide('route', route)

Its also still not resolved in the files as seen below:
![image](https://user-images.githubusercontent.com/39704356/179360188-6cd7acfd-e532-4ffe-b16c-cae32c37a7b6.png)

**Reply by romsar** [reply]
Above your loadCurrencies func put:
```
const route = inject('route')
```

Don't forget to import `inject` from `vue`.

> How come we removed the resolvecomponent part?

My code is pretty old (~1 year), so don't change your code except:
```
const Ziggy = props.initialPage.props.ziggy;

.mixin({
        methods: {
          route: (name, params, absolute, config = Ziggy) => route(name, params, absolute, config),
        },
      })
      .provide('route', (name, params, absolute, config = Ziggy) => route(name, params, absolute, config))

```
